Fix command line vs. native file precedence error on regen

blue42u opened the pull request:
Fixes https://github.com/mesonbuild/meson/issues/11930

The [logic in `Environment.__init__`](https://github.com/mesonbuild/meson/blob/9a6a95483c4d3c4f91824f1a7e52fc781e111043/mesonbuild/environment.py#L479-L498) only loads the options specified from the `setup` command line if the `CoreData` failed to load. This means on regeneration (`setup --reconfigure`) the options as seen by `get_option()` will not include the effect of command line options.

This PR makes the call to `read_cmd_line_file` non-conditional to fix the issue. `test_user_options_command_line_overrides` is extended to cover this case.

**ADD:** This PR also makes the application of `default_options` similarly non-conditional. Note that this means changing `default_options` now will change the option during the following reconfigure (if it wasn't set on the command line`). This is a deviation from the behavior of 1.1.1 and prior.

blue42u committed:
Extend unit tests to test reconfiguration

`NativeFileTests::test_user_options_command_line_overrides` only checks
that the initial `meson setup` works, it does not check that the state
is adequately preserved in a later `meson setup --reconfigure`. Add an
extra call for this case.

blue42u committed:
Always restore options from the cmd_line_file

In `Environment.__init__`, if the CoreData exists and loads successfully
the original options passed to `meson setup` are not restored by
`read_cmd_line_file`. This causes inconsistencies between the
`CoreData.options` and `Environment.options`.

Make this call unconditional so that these two `options` are consistent.

Fixes https://github.com/mesonbuild/meson/issues/11930

xclaesse commented:
This is an easy workaround indeed, would have to see why CI did not pass.

I've been wanting to implement a more complete fix for this for a while. The idea is we should store in coredata the "source" of an option value (default_options kwarg, env, cross file, cli, etc) and only override a value when it's coming from a higher priority source. This would allow for example to update an option when `default_options` kwarg is changed if the current value was not from a higher priority source. That's how Muon (a C implementation of Meson) works.

blue42u committed:
Always apply the default_options

In the implementation of `project()`, the `default_options` are only
applied on the first invocation and not on later ones. This seems to
have been a workaround to avoid disturbing the command line options on
reconfigure (which weren't actually present, so...? Confusing.)

Make this application unconditional so any options set by
`default_options` are still set after a reconfigure.

Note that this means changes to `default_options` *will* take effect
after a reconfiguration (if not previously overridden on the command
line). This is a user-visible change in behavior from 1.1.1 and prior.

blue42u commented:
I think I've finally figured it out, there was a conditional in `Interpreter.func_project` that was preventing `default_options` from being applied at the right time. Workarounds for workarounds. :sweat_smile:

Commits added and rebased, and OP updated.

> The idea is we should store in coredata the "source" of an option value...

I really like this idea. If I understood the current code better I would try to implement it, it seems like it would fix a lot of option-related edge cases.

Would you happen to know why `CoreData.options` and `Environment.options` are handled separately? AFAIU there isn't any substantial difference between them. I think like removing that duplication would significantly uncomplicate the situation so as it is. :thinking:

tristan957 commented:
targeted for 1.3.0 @xclaesse
